### PR TITLE
[C/PyTorch] Add THD support for cuDNN attention

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

This PR adds THD support for fused attention (`F16_arbitrary_seqlen` backend). This feature allows users to run attention for two more cases:
```
case 1: no padding between sequences
    8 sequences are packed as abcddddeeeffgh
    total sequence length t = 14
    cumulative sequence lengths: cu_seqlens = [0, 1, 2, 3, 7, 10, 12, 13, 14]
    offset tensor that helps cuDNN step into qkv tensors: offset_tensor = [0, 1, 2, 3, 7, 10, 12, 13]

case 2: with padding between sequences
    8 sequences are packed as abc0ddddeee0ffgh
    total sequence length t = 16
    cumulative sequence lengths: cu_seqlens = [0, 1, 2, 3, 7, 10, 12, 13, 14]
    offset tensor that helps cuDNN step into qkv tensors: offset_tensor = [0, 1, 2, 4, 8, 12, 14, 15]
```

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Changes

- Added THD support to cuDNN attention on the C level and in PyTorch modules. Jax and Paddle are to be worked on.
- Updated cudnn-frontend to v1.4.0.
- Added tests for `max_seqlen_q=1` and `head_dim=256` inference cases for `thd`, `bshd`, and `sbhd` formats.

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by zlsh80826** [comment]
This PR breaks the fused attention API. I'm wondering if we can change `cu_seqlens` arguments to `actual_seqlens`, so that we can avoid the kernel to convert `actual_seqlens` to `cu_seqlens`.
Actually, `cu_seqlens` as API makes

**Comment by zlsh80826** [comment]
I had a [commit](https://github.com/zlsh80826/TransformerEngine/commit/74bf8a27d165c44bf87e14a69a551ca1353e25fb) to fix the JAX build issue with the new API.

**Comment by cyanguwa** [comment]
> This PR changes the fused attention API. I'm wondering if we can also change `cu_seqlens` arguments to `actual_seqlens`, so that we can avoid the kernel to convert `actual_seqlens` to `cu_seqlens`.
> 
> Actually, `cu_seqlens` as API arguments makes JAX do two additional kernels, which is `mask -> actual_seqlens -> sharding (DP) -> kernel to convert actual_seqlens to cu_seqlens -> call nvte_fmha -> kernel to convert cu_seqlens to actual_seqlens`.

@zlsh80826 If `mask -> cu_seqlens` instead of `mask -> actual_seqlens`, would there still be two extra kernels? @ptrendx suggested that we make the `cu_seqlens/actual_seqlens` related changes in a different PR in order to keep this PR focused on THD. Any burning needs to make the `cu_seqlens/actual_seqlens` changes?

**Comment by zlsh80826** [comment]
> @zlsh80826 If `mask -> cu_seqlens` instead of `mask -> actual_seqlens`, would there still be two extra kernels? @ptrendx suggested that we make the `cu_seqlens/actual_seqlens` related changes in a different PR in order to keep this PR focused on THD. Any burning needs to make the `cu_seqlens/actual_seqlens` changes?

If changes to `mask->actual_seqlens` then it will not have two extra kernels. I did profile the performance impact and the extra kenrels are less then 0.1% in the e2e training, so it is not hurry for now.

**Comment by zlsh80826** [comment]
Hi @cyanguwa, the changes for common looks good for me. I will fix the JAX build this week.

File: transformer_engine/common/include/transformer_engine/fused_attn.h [hunk 1]
```diff
@@ -176,7 +176,11 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
  *  \param[out]    O                        The output O tensor.
  *  \param[out]    Aux_CTX_Tensors          Auxiliary output tensors when training,
  *                                          e.g. M, ZInv, rng_state.
- *  \param[in]     cu_seqlens               Accumulative sequence lengths, [batch_size + 1].
+ *  \param[in]     cu_seqlens               Cumulative sequence lengths, [batch_size + 1].
+ *  \param[in]     seq_offsets_q            Cumulative sequence offsets for Q, [batch_size + 1].
+ *  \param[in]     seq_offsets_k            Cumulative sequence offsets for K, [batch_size + 1].
+ *  \param[in]     seq_offsets_v            Cumulative sequence offsets for V, [batch_size + 1].
+ *  \param[in]     seq_offsets_o            Cumulative sequence offsets for O, [batch_size + 1].
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/include/transformer_engine/fused_attn.h:201]: Done.

File: transformer_engine/common/fused_attn/fused_attn.cpp [hunk 2]
```diff
@@ -135,20 +135,24 @@ NVTE_Fused_Attn_Backend nvte_get_fused_attn_backend(
     }
     if (((cudnn_runtime_version >= 8903 && sm_arch_ >= 80)
                 || (cudnn_runtime_version < 8903 && (sm_arch_ == 80 || sm_arch_ == 90)))
-            && (max_seqlen_q % 64 == 0)
-            && (max_seqlen_kv % 64 == 0)
+            && ((cudnn_runtime_version < 90000 && max_seqlen_q % 64 == 0 && max_seqlen_kv % 64 == 0)
+                || (cudnn_runtime_version >= 90000))
```
Comment (zlsh80826) [inline transformer_engine/common/fused_attn/fused_attn.cpp:139]: Does cuDNN 9.0 support all seqlen? I feed max_seqlen == 32 and it seems not supported.

**Review by zlsh80826 — APPROVED** [review]
LGTM

**Comment by zlsh80826** [comment]
@cyanguwa I have a few questions when exposing the THD format on JAX. Could you take a look on cudnn slack channel before merging it? Thanks.

**Pull request merged**